src/ripple/basics/Number.h

```diff
@@ -0,0 +1,340 @@
+    constexpr static std::int64_t minMantissa = 1'000'000'000'000'000LL;
+    constexpr static std::int64_t maxMantissa = 9'999'999'999'999'999LL;
+
+    // The range for the exponent when normalized
+    constexpr static int minExponent = -32768;
+    constexpr static int maxExponent = 32768;
+
+    class Guard;
+};
+
+inline constexpr Number::Number(rep mantissa, int exponent, unchecked) noexcept
+    : mantissa_{mantissa}, exponent_{exponent}
+{
+}
+
+inline Number::Number(rep mantissa, int exponent)
+    : mantissa_{mantissa}, exponent_{exponent}
+{
+    normalize();
+}
+
+inline Number::Number(rep mantissa) : Number{mantissa, 0}
+{
+}
+
+inline Number::Number(IOUAmount const& x) : Number{x.mantissa(), x.exponent()}
+{
+}
+
+inline Number::Number(XRPAmount const& x) : Number{x.drops()}
+{
+}
+
+inline constexpr Number::rep
+Number::mantissa() const noexcept
+{
+    return mantissa_;
+}
+
+inline constexpr int
+Number::exponent() const noexcept
+{
+    return exponent_;
+}
+
+inline constexpr Number
+Number::operator+() const noexcept
+{
+    return *this;
+}
+
+inline constexpr Number
+Number::operator-() const noexcept
+{
+    auto x = *this;
+    x.mantissa_ = -x.mantissa_;
+    return x;
+}
+
+inline Number&
+Number::operator++()
+{
+    *this += Number{1000000000000000, -15, unchecked{}};
+    return *this;
+}
+
+inline Number
+Number::operator++(int)
+{
+    auto x = *this;
+    ++(*this);
+    return x;
+}
+
+inline Number&
+Number::operator--()
+{
+    *this -= Number{1000000000000000, -15, unchecked{}};
+    return *this;
+}
+
+inline Number
+Number::operator--(int)
+{
+    auto x = *this;
+    --(*this);
+    return x;
+}
+
+inline Number&
+Number::operator-=(Number const& x)
+{
+    return *this += -x;
+}
+
+inline Number
+operator+(Number const& x, Number const& y)
+{
+    auto z = x;
+    z += y;
+    return z;
+}
+
+inline Number
+operator-(Number const& x, Number const& y)
+{
+    auto z = x;
+    z -= y;
+    return z;
+}
+
+inline Number
+operator*(Number const& x, Number const& y)
+{
+    auto z = x;
+    z *= y;
+    return z;
+}
+
+inline Number
+operator/(Number const& x, Number const& y)
+{
+    auto z = x;
+    z /= y;
+    return z;
+}
+
+inline Number::operator IOUAmount() const
+{
+    return IOUAmount{mantissa(), exponent()};
+}
+
+inline constexpr bool
+Number::isnormal() const noexcept
+{
+    auto const abs_m = mantissa_ < 0 ? -mantissa_ : mantissa_;
+    return minMantissa <= abs_m && abs_m <= maxMantissa &&
+        minExponent <= exponent_ && exponent_ <= maxExponent;
+}
+
+inline constexpr Number
+abs(Number x) noexcept
+{
+    if (x < Number{})
+        x = -x;
+    return x;
+}
+
+// Returns f^n
+// Uses a log_2(n) number of multiplications
+
+Number
+power(Number f, unsigned n);
+
+// Returns f^(1/d)
+// Uses Newton–Raphson iterations until the result stops changing
+// to find the root of the polynomial g(x) = x^d - f
+
+Number
+root(Number f, unsigned d);
```

Comment:
The implementation of `root` modifies the passed in `f`, so there's a good reason to be passing a non-ref `Number` here.

Reply:
Before we add these functions, I'd like to make sure they are needed. We need to be able to take the square root of a number, and we need to be able to square a number (multiply is likely good enough). I'd like to not add `root` and ``pow` until we know we need them for sure.

Reply:
AMM equations require at a minimum `root` in case of equal weights. `pow` is required if unequal weights is supported.

Reply:
@gregtatcam Don't we only need square root, not a general root function?

Reply:
Correct, just the `square root`.

Reply:
For the unequal weights case (if we support that), we need the general root function to support the power raise to a general rational function.